index.js

```diff
@@ -16,17 +16,28 @@ const moduleSupport = semver.satisfies(process.version, '>= 14 || >= 12.17.0 < 1
 const defaults = {
   scriptPattern: /((^.?|\.[^d]|[^.]d|[^.][^d])\.ts|\.js|\.cjs|\.mjs)$/i,
   indexPattern: /^index(\.ts|\.js|\.cjs|\.mjs)$/i,
+  autoHooksPattern: /^[_.]?auto_?hooks(\.ts|\.js|\.cjs|\.mjs)$/i,
   dirNameRoutePrefix: true
 }
 
+let fp = null
+
 const fastifyAutoload = async function autoload (fastify, options) {
   const packageType = await getPackageType(options.dir)
   const opts = { ...defaults, packageType, ...options }
   const plugins = await findPlugins(opts.dir, opts)
   const pluginsMeta = {}
 
-  await Promise.all(plugins.map(({ file, type, prefix }) => {
-    return loadPlugin(file, type, prefix, opts)
+  try {
+    // bring in fp so we can expose the hooks to the parent plugin
+    if (opts.autoHooks) fp = require('fastify-plugin')
```

Comment:
just require it at the top. Anyway, it's not strictly needed. The only neeeded is https://github.com/fastify/fastify-plugin/blob/52cec31e4c0dc4eb6ecc2b2f0e2f1c21994713d1/plugin.js#L21.

Reply:
> 
> 
> Looking at the code now, it might be simpler to leverage https://github.com/fastify/fastify/blob/master/docs/Encapsulation.md and add an intermediate parent when the hooks are present.

I think this is how I handled it. When there are hooks, it will create an intermediate parent (`fastify.register...`) and register the hook(s) and plugin within that scope (`app.register...`). Otherwise it just registers the plugin at the `fastify.register` level. Did I get some syntax wrong?

> This could remove some of those options and have a simpler behavior.
> 
> Wdyt?
> 

I agree, the first draft of the examples in the documentation make it feel confusing. 

My reasoning behind the two additional hook attachment options was that it felt most logical to cascade by default as that's the way folder permissions tend to work in most operating systems (i.e. if you own a parent folder, and you create subfolders, the ownership cascades until you override it). Same with the default ``.htaccess``-style basic auth (i.e. the parent folder and all the children require the same username and password until you override it with a new .htaccess file)

That would be useful for cases like a news site with a paywall that has different membership levels, or a multi-tenant application with differing feature tiers.

E.g.

  ```
  ├── routes
  │   ├── paywall
  │   │   ├── .autohooks.js // check membership
  │   │   ├── routes.js
  │   │   └── archive
  │   │       └── archive-routes.js // membership check inherited from parent
  │   │           └── premium_archive
  │   │               ├── .autohooks.js // premium membership check
  │   │               └── routes.js
  │   └── public_routes.js
  └── app.js
  ```

But I could also see situations where just one or two folders inside a tree could need a set of hooks (e.g. cache headers).

Maybe I could rename the options to make them more logical?

> just require it at the top. Anyway, it's not strictly needed. 

Done. Much cleaner!

> The only neeeded is https://github.com/fastify/fastify-plugin/blob/52cec31e4c0dc4eb6ecc2b2f0e2f1c21994713d1/plugin.js#L21.

The main Fastify `Plugin.md` suggested that using fp was the best practice to future-proof the plugin. I wasn't excited about adding a dependency, but I didn't want to add anything that might break in a core update down the road.

> (This also needs tests).

For sure. I started my first project using Fastify last week (and absolutely love the flexibility and speed) so I'm still getting familiar with the code.